biz.aQute.bndlib/src/aQute/bnd/osgi/PermissionGenerator.java

```diff
@@ -0,0 +1,145 @@
+			} else if (KEY.equals(arg))
+				;
+			else
+				builder.error("Invalid option in ${permissions}: %s", arg);
+		}
+
+		return sb.toString();
+	}
+
+	private void generateAdmin() {
+		sb.append("(org.osgi.framework.AdminPermission)");
+	}
+
+	private void generateCapability() {
+		for (String namespace : builder.getProvideCapability().keySet()) {
+			if (!Processor.isDuplicate(namespace)) {
+				sb.append("(org.osgi.framework.CapabilityPermission \"").append(namespace)
+						.append("\" \"provide\")\r\n");
+			}
+		}
+		for (String namespace : builder.getRequireCapability().keySet()) {
+			if (!Processor.isDuplicate(namespace)) {
+				sb.append("(org.osgi.framework.CapabilityPermission \"").append(namespace)
+						.append("\" \"require\")\r\n");
+			}
+		}
+	}
+
+	private void generatePackages() {
+		if (builder.getImports() != null) {
+			for (PackageRef imp : builder.getImports().keySet()) {
+				if (!imp.isJava()) {
+					sb.append("(org.osgi.framework.PackagePermission \"");
+					sb.append(imp);
+					sb.append("\" \"import\")\r\n");
+				}
+			}
+		}
+		if (builder.getExports() != null) {
+			for (PackageRef exp : builder.getExports().keySet()) {
+				sb.append("(org.osgi.framework.PackagePermission \"");
+				sb.append(exp);
+				sb.append("\" \"export\")\r\n");
+			}
+		}
+	}
+
+	private void generateServices() {
+		for (String declaredService : getDeclaredServices()) {
+			sb.append("(org.osgi.framework.ServicePermission \"").append(declaredService)
+					.append("\" \"register\")\r\n");
+		}
+		for (String referencedService : getReferencedServices()) {
+			sb.append("(org.osgi.framework.ServicePermission \"").append(referencedService).append("\" \"get\")\r\n");
+		}
+	}
+
+	public Set<String> getDeclaredServices() {
+		Set<String> declaredServices = new TreeSet<>();
+		for (Entry<String,Attrs> entry : builder.getProvideCapability().entrySet()) {
+			if (Processor.removeDuplicateMarker(entry.getKey()).equals("osgi.service")) {
+				Attrs attrs = entry.getValue();
+				String ifaces = attrs.get("objectClass");
+				if (ifaces != null) {
+					for (String iface : ifaces.split(",")) {
+						declaredServices.add(iface);
+					}
+				}
+			}
+		}
+		return declaredServices;
+	}
+
+	private static final Pattern	FILTER_PATTERN	= Pattern
+															.compile("^\\(objectClass=([\\p{javaJavaIdentifierPart}\\.]+)\\)$");
+
+	public Set<String> getReferencedServices() {
+		Set<String> referencedServices = new TreeSet<>();
+		for (Entry<String,Attrs> entry : builder.getRequireCapability().entrySet()) {
+			if (Processor.removeDuplicateMarker(entry.getKey()).equals("osgi.service")) {
+				Attrs attrs = entry.getValue();
+					String filter = attrs.get("filter:");
+					if (filter != null && !filter.isEmpty()) {
+					Matcher matcher = FILTER_PATTERN.matcher(filter);
+					if (matcher.matches()) {
+						referencedServices.add(matcher.group(1));
+					} else {
+						builder.warning("Unexpected osgi.service requirement filter: {}", filter);
```

Comment:
A user could specify a valid filter directive without an objectClass. So we should not warn here. Just ignore.